File: backend/src/main/java/django/TimestampSigner.java
package django;

import javax.annotation.Nullable;
import java.time.Duration;
import java.time.Instant;

public class TimestampSigner extends Signer {
    public TimestampSigner(String key, String salt) {
        super(key, salt);
    }

    @Override public String unsign(String signedValue) {
        return unsign(signedValue, null);
    }

    public String unsign(String signedValue, @Nullable Duration maxAge) {
        String result = super.unsign(signedValue);
        String[] parts = result.split(SEPARATOR, 2);
        String value = parts[0];
        Instant timestamp = Instant.ofEpochSecond(BaseConverter.BASE62.decode(parts[1]));
        if (maxAge != null) {
            // Check timestamp is not older than max_age
            Instant expiry = timestamp.plus(maxAge);
            if (Instant.now().isAfter(expiry)) {
                throw new SignatureExpired("signature expired");
            }
        }
        return value;
    }
}

File: backend-ts/src/auth/django/timestamp_signer.ts
import { BaseConverter } from "./base_converter";
import { Signer } from "./signer";

export class SignatureExpired extends Error {}

export class TimestampSigner extends Signer {
  public sign(value: string): string {
    value = value + Signer.SEPARATOR + this.timestamp()
    return super.sign(value)
  }

  private timestamp(): string {
    return BaseConverter.BASE62.encode(Date.now() / 1000)
  }

  public unsign(signedValue: string, maxAgeMs?: number): string {
    const result = super.unsign(signedValue)
    const parts = result.split(Signer.SEPARATOR, 2)
    const value = parts[0]
    const timestamp = BaseConverter.BASE62.decode(parts[1]) * 1000
    if (maxAgeMs != null) {
      // Check timestamp is not older than max_age
      const expiry = timestamp + maxAgeMs
      if (Date.now() > expiry) {
        throw new SignatureExpired("signature expired")
      }
    }
    return value;
  }
}
